kyo-kernel/shared/src/main/scala/kyo/kernel/Isolate.scala

```diff
@@ -0,0 +1,370 @@
+        def apply[R, P](using s: Contextual[R, P]): Contextual[R, P] = s
+
+        /** Derives a Contextual isolate instance based on available instances. */
+        inline given derive[R, P]: Contextual[R, P] = ${ deriveImpl[R, P] }
+
+        private def deriveImpl[R: Type, P: Type](using Quotes): Expr[Isolate.Contextual[R, P]] =
+            commonDeriveImpl[R, P, Isolate.Contextual[R, P]] { (retain, _) =>
+                import quotes.reflect.*
+
+                val missing = retain.filter { t =>
+                    t.asType match
+                        case '[tpe] => Expr.summon[Isolate.Contextual[tpe, P]].isEmpty
+                }
+
+                if missing.nonEmpty then
+                    report.errorAndAbort(
+                        s"""|This operation requires Contextual isolation for effects:
+                            |
+                            |  ${missing.map(_.show.red).mkString(" & ")}
+                            |
+                            |Common mistake: Using operations like Async.run with effects that need complex state management.
+                            |
+                            |You have a couple of options, from simplest to most advanced:
+                            |
+                            |1. Handle these effects before the operation:
+                            |   Async.run(MyEffect.run(computation))
+                            |
+                            |2. Use an operation that supports complex state:
+                            |   Instead of Async.run, use Async.parallel 
+                            |""".stripMargin
+                    )
+                end if
+
+                '{ cached.asInstanceOf[Isolate.Contextual[R, P]] }
+            }
+
+    end Contextual
+
+    /** Controls state isolation for effects requiring structured state management.
+      *
+      * Provides a three-phase approach to state isolation:
+      *   1. Capture initial state before forking
+      *   2. Transform state during isolated execution
+      *   3. Restore transformed state when the fork completes
+      *
+      * @tparam Retain
+      *   The effects that will be satisfied by this isolation
+      * @tparam Passthrough
+      *   Additional effects that will remain pending after isolation
+      */
+    abstract class Stateful[Retain, -Passthrough]:
+        self =>
+
+        /** The type of state being managed */
+        type State
+
+        /** How state is transformed during isolated execution */
+        type Transform[_]
+
+        /** Evidence that transformed state can be properly handled */
+        given flatTransform[A: Flat]: Flat[Transform[A]]
+
+        /** Runs a computation with full state lifecycle management.
+          *
+          * Convenience method that composes capture, isolate and restore to handle the complete state lifecycle.
+          */
+        def run[A: Flat, S](v: A < (S & Retain))(using Frame): A < (S & Retain & Passthrough) =
```

Comment:
Should this be final?
```suggestion
        final def run[A: Flat, S](v: A < (S & Retain))(using Frame): A < (S & Retain & Passthrough) =
```